test/FunctionalTests/Microsoft.OData.Core.Tests/ODataNotificationReaderTests.cs

```diff
@@ -0,0 +1,161 @@
+            // We care about the notification reader being disposed
+            // We don't care about the reader passed to the notification reader
+            using (var notificationReader = new ODataNotificationReader(
+                this.reader,
+                this.streamListener,
+                synchronous))
+            {
+            }
+
+            var result = ReadStreamContents();
+
+            Assert.Equal(expected, result);
+        }
+
+        [Theory]
+        [InlineData(true, "StreamDisposed")]
+        [InlineData(false, "StreamDisposedAsync")]
+        public void NotificationReaderDisposeShouldBeIdempotent(bool synchronous, string expected)
+        {
+            var notificationReader = new ODataNotificationReader(
+                this.reader,
+                this.streamListener,
+                synchronous);
+
+            // 1st call to Dispose
+            notificationReader.Dispose();
+            // 2nd call to Dispose
+            notificationReader.Dispose();
+
+            var result = ReadStreamContents();
+
+            // StreamDisposed/StreamDisposeAsync was written only once
+            Assert.Equal(expected, result);
+        }
+
+#if NETCOREAPP3_1
+        [Fact]
+        public async Task NotificationReaderDisposeShouldInvokeStreamDisposedAsync()
+        {
+            await using (var notificationReader = new ODataNotificationReader(
+                this.reader,
+                this.streamListener)) // `synchronous` argument becomes irrelevant since we'll directly call DisposeAsync
+            {
+            }
+
+            var result = await this.ReadStreamContentsAsync();
+
+            Assert.Equal("StreamDisposedAsync", result);
+        }
+
+        [Fact]
+        public async Task NotificationReaderDisposeAsyncShouldBeIdempotent()
+        {
+            var notificationReader = new ODataNotificationReader(
+                this.reader,
+                this.streamListener);
+
+            // 1st call to DisposeAsync
+            await notificationReader.DisposeAsync();
+            // 2nd call to DisposeAsync
+            await notificationReader.DisposeAsync();
+
+            var result = await this.ReadStreamContentsAsync();
+
+            // StreamDisposeAsync was written only once
+            Assert.Equal("StreamDisposedAsync", result);
+        }
+
+#else
+        [Fact]
+        public async Task NotificationReaderDisposeShouldInvokeStreamDisposedAsync()
+        {
+            using (var notificationReader = new ODataNotificationReader(
+                this.reader,
+                this.streamListener,
+                /*synchronous*/ false))
+            {
+            }
+
+            var result = await this.ReadStreamContentsAsync();
+
+            Assert.Equal("StreamDisposedAsync", result);
+        }
+#endif
+
+        private string ReadStreamContents()
+        {
+            this.stream.Position = 0;
+            return new StreamReader(this.stream).ReadToEnd();

```

Comment:
Please make sure to dispose this intermediate object before the reference to it is lost

Reply:
Fixed by having the test class implement `IDisposable` interface and implementing the `Dispose` method that Xunit relies on for cleanup after each test is ran